src/asciifile.rs

```diff
@@ -16,22 +18,62 @@ pub enum EncodingError {
 
 const ENCODING_ERROR_MAX_CONTEXT_LEN: usize = 180;
 
+//enum LineContext {
+//Truncated,
+//NotTruncated
+//}
+
 impl<'m> AsciiFile<'m> {
+    /// return the start index of the line without the newline character (\n).
+    /// This means you can make slices that do NOT include the newline
+    /// character by simply using the return value of this function as
+    /// lower bound X in the range X..Y.
+    ///
+    /// If the cursor/byte offset is on a newline character, the newline
+    /// character will belong to the previous line, meaning the return
+    /// value will be the given byte offset and
+    /// not the byte offset of the next new line.
+    fn get_line_end_idx(mapping: &'m [u8], byte_offset: usize) -> usize {
+        let region_end = mapping
+            .len()
+            .min(byte_offset + ENCODING_ERROR_MAX_CONTEXT_LEN);
+
+        let region = &mapping[byte_offset..region_end];
+        region
+            .iter()
+            .position(|&chr| chr == '\n' as u8)
+            .map(|pos| pos + byte_offset)
+            .unwrap_or(region_end)
+    }
+
+    /// return the end index of the line includes the newline character (\n).
+    /// This means you can make slices that do NOT include the newline
+    /// character by simply using the return value of this function as
+    /// upper bound Y in the range X..Y.
+    ///
+    /// If the cursor/byte offset is on a newline character, the newline
+    /// character will belong to the previous line, meaning the return
+    /// value will not be the given by the byte
+    /// offset, but the byte offset of the previous new line.
+    fn get_line_start_idx(mapping: &'m [u8], byte_offset: usize) -> usize {
+        let region_start = byte_offset
+            .checked_sub(ENCODING_ERROR_MAX_CONTEXT_LEN)
+            .unwrap_or(0);
+
+        let region = &mapping[region_start..byte_offset];
```

Comment:
This could panic if `byte_offset > mapping.len()`. The same could happen in the function `get_line_end_idx` above.

Reply:
Fixed in commit: [reiner-diagnostics f733f22] add debug_assert to byte_offset methods